### DISCUSSION TITLE
Remove superfluous const parameters from function declarations

### DISCUSSION THREAD

**Discussion by JuanDiegoMontoya** [GraphicsProgramming/learnd3d11#106] [General]
For non-reference, non-pointer function prototype parameters, `const` is superfluous and has no effect. It only serves to clutter the API and does not provide any guarantee to the caller. See [this SO answer](https://stackoverflow.com/a/11036371) for more info.

Suggestion: remove `const` from non-reference, non-pointer-like parameters in function declarations, and keep them in function *definitions* to maintain const-correctness where it matters, for example:
```cpp
class DearImGuiApplication final : public Application
{
public:
    ...
    void OnResize(
        int32_t width,
        int32_t height) override;
    ...
};

void DearImGuiApplication::OnResize(
    const int32_t width,
    const int32_t height)
{
    ...
}
```

The other place I have noticed this issue is with the `std::string_view` parameter of `Application`'s (and its derivatives') constructor.

**Comment by deccer** [comment]
As discussed in discord. 
I would like to keep const in hpp/cpp to be consistent. Or does it absolutely not make sense at all (i am not really convinced by the attached SO, as in... keeping it shows intent, and doesn't really "pollute" anything)?

**Comment by JuanDiegoMontoya** [comment]
If consistency is the goal, then at least these offest parameters should be changed in `DeviceContext` (to be `const`):
```cpp
    void SetVertexBuffer(
        ID3D11Buffer* vertexBuffer,
        uint32_t vertexOffset);
    void SetIndexBuffer(
        ID3D11Buffer* indexBuffer,
        uint32_t indexOffset);
```

**Reply by deccer** [reply]
Yes definitlylelye all "missing consts" should be added accordingly

**Reply by deccer** [reply]
Probably includes `void Foo() const;` those kind of consts as well, I  am sure i didn't add them where possible either and left some methods out ;)

**Comment by spnda** [comment]
clang-tidy has the same opinion on this issue. As non-reference and non-pointer arguments are passed by value and the caller therefore doesn't need to know whether the function implementation restricts the copied value to be const, it only makes sense to note the const declaration within the function definition, which usually will reside in the .cpp TU. The only point at which it is relevant knowing whether the value is const or not is within the function definition, and nowhere else is this information relevant. If you're writing an API you are focusing on writing within the source file, and the header file is only the interface to the outside and as they're not interested and have no reason to know whether it's const there is no reason for it to be included there. The header can be freed from any const, making the header easier to read while keeping the API functionally identical.

I don't think this should be ignored simply for consistency, as it is a extra burden for writing and reading while bringing absolutely nothing to the table. Additionally, the header could be marked const while the definition isn't and the C++ standard didn't do this for no reason. It would also be possible  to just add the definition to the header and mark it inline, if you want the const-ness to be visible to any callers.

**Comment by deccer** [comment]
Ok then lets remove const from hpp where applicable
